src/highdicom/seg/sop.py

```diff
@@ -1523,29 +1591,49 @@ def __init__(
 
                 # Add the segmentation pixel array for this frame to the list
                 if is_encaps:
-                    # Encode this frame and add to the list for encapsulation
-                    # at the end
-                    full_frames_list.append(segment_array[plane_index])
+                    if process_pool is None:
+                        # Encode this frame and add resulting bytes to the list
+                        # for encapsulation at the end
+                        compressed_frames_list.append(
+                            encode_frame(
+                                segment_array,
+                                **encode_frame_kwargs,
+                            )
+                        )
+                    else:
+                        # Submit this frame for encoding this frame and add the
+                        # future to the list for encapsulation at the end
+                        future = process_pool.submit(
+                            encode_frame,
+                            array=segment_array,
+                            **encode_frame_kwargs,
+                        )
+                        frame_futures_list.append(future)
                 else:
-                    # Concatenate the 1D array for re-encoding at the end
-                    full_frames_list.append(
-                        segment_array[plane_index].flatten()
-                    )
+                    # Concatenate the 1D array for encoding at the end
+                    full_frames_list.append(segment_array.flatten())
```

Comment:
I wonder whether a separate variable is needed or whether frames (compressed or not) could just be stored in a list called `frames`? How the list of frames then needs to be handled (encapsulation or native encoding) is determined by `is_enaps`.

The variable for the futures could be called `frames_futures`.

In my opinion, that would make it easier to follow.

Reply:
I did it this way primarily to appease mypy, but I guess it probably hurts readability. Implemented 5c5cbb6 as you suggest